src/combinatorics/set_family/compression/uv.lean

```diff
@@ -0,0 +1,256 @@
+/-
+Copyright (c) 2021 Bhavik Mehta. All rights reserved.
+Released under Apache 2.0 license as described in the file LICENSE.
+Authors: Yaël Dillies, Bhavik Mehta
+-/
+import data.finset.basic
+
+/-!
+# UV-compressions
+
+This file defines UV-compression.
+
+A compression in combinatorics is an operation on a set family that reduces its shadow.
+UV-compressing `a : α` along `u v : α` means replacing `a` by `(a ⊔ u) \ v` if `a` and `u` are
+disjoint and `v ≤ a`. In some sense, it's moving `a` from `v` to `u`.
+
+UV-compressions are immensely useful to prove the Kruskal-Katona theorem. The idea is that
+compressing a set family might decrease the size of its shadow, so iterated compressions hopefully
+minimise the shadow.
+
+## Main declarations
+
+* `uv.compress`: `compress u v a` is `a` compressed along `u` and `v`.
+* `uv.compress_motion`: `compress_motion u v s` is the part of `s` that can move along the
+  compression. Precisely, it's the compression of the elements of `s` whose compression is not
+  already in `s`.
+* `uv.compress_remains`: `compress_remains u v s` is the part of `s` that can't move along the
+  compression. Precisely, it's the element whose compression is already in `s`.
+* `uv.compression`: `compression u v s` is the compression of the set family `s` along `u` and `v`.
+  It is the union of `compress_motion u v s` and `compress_remains u v s`. This way of splitting
+  into what moves and what does not ensures the compression doesn't squash the set family, which is
+  proved by `uv.card_compress`.
+
+## Notation
+
+`𝓒` is notation for `uv.compression` in locale `finset_family`.
+
+## Notes
+
+Even though our emphasis is on `finset α`, we define UV-compressions more generally in a generalized
+boolean algebra, so that one can use it for `set α`.
+
+## TODO
+
+Prove that compressing reduces the shadow. This result and some more already exist on the branch
+`combinatorics`.
+
+## References
+
+* https://github.com/b-mehta/maths-notes/blob/master/iii/mich/combinatorics.pdf
+
+## Tags
+
+compression, UV-compression, shadow
+-/
+
+open finset
+
+variable {α : Type*}
+
+-- The namespace here is useful to distinguish between other compressions.
+namespace uv
+
+/-! ### UV-compression in generalized boolean algebras -/
+
+section generalized_boolean_algebra
+variables [generalized_boolean_algebra α] [decidable_rel (@disjoint α _)]
+  [decidable_rel ((≤) : α → α → Prop)] {s : finset α} {u v a b : α}
+
+/-- To UV-compress `A`, if it doesn't touch `U` and does contain `V`, we remove `V` and
+put `U` in. We'll only really use this when `|U| = |V|` and `U ∩ V = ∅`. -/
+def compress (u v a : α) : α := if disjoint u a ∧ v ≤ a then (a ⊔ u) \ v else a
+
+/-- Part of the compressed family, where we move the sets whose compression is not there. -/
+@[reducible]
```

Comment:
I don't like that these things are reducible. Does it work to remove this attribute and make the `mem_` lemmas below `simp` or just to use them more?

Reply:
Oh yeah I just kept that from your original branch. Better leave them semireducible.